File: rust/src/bin/2.rs
use std::collections::HashMap;
use util::read_lines;

fn main() {
    let first_col_map = HashMap::from([('A', 'R'), ('B', 'P'), ('C', 'S')]);
    let second_col_map = HashMap::from([('X', 'R'), ('Y', 'P'), ('Z', 'S')]);
    let value = HashMap::from([('R', 1), ('P', 2), ('S', 3)]);
    let defeats = HashMap::from([('R', 'S'), ('P', 'R'), ('S', 'P')]);
    let mut defeated_by = HashMap::<char, char>::new();
    for (k, v) in &defeats {
        defeated_by.insert(*v, *k);
    }

    if let Ok(lines) = read_lines("2.txt") {
        let mut score_1 = 0u32;
        let mut score_2 = 0u32;

        for line in lines {
            if let Ok(line_str) = line {
                let bytes = line_str.as_bytes();

                let first = bytes[0] as char;
                let second = bytes[bytes.len() - 1] as char;

                let them = first_col_map[&first];
                let me = second_col_map[&second];

                score_1 += value[&me];
                if me == them {
                    score_1 += 3
                } else if defeats[&me] == them {
                    score_1 += 6
                }

                if second == 'Y' {
                    score_2 += 3 + value[&them];
                } else if second == 'X' {
                    score_2 += value[&defeats[&them]];
                } else {
                    score_2 += 6 + value[&defeated_by[&them]];
                }
            }
        }

        println!("{} {}", score_1, score_2);
    }
}

File: python/2.py
first_col_map = {
    'A': 'R',
    'B': 'P',
    'C': 'S'
}

second_col_map = {
    'X': 'R',
    'Y': 'P',
    'Z': 'S'
}

val = {
    'R': 1,
    'P': 2,
    'S': 3,
}

defeats = {
    'R': 'S',
    'S': 'P',
    'P': 'R',
}

defeated_by = {v: k for k, v in defeats.items()}

with open('../data/2.txt', 'r') as f:
    score = 0
    score_2 = 0
    data = f.readlines()
    for line in data:
        a, b = line[0], line[-2]

        them = first_col_map[a]
        me = second_col_map[b]

        score += val[me]
        if me == them:
            score += 3
        elif defeats[me] == them:
            score += 6

        if b == 'Y':
            score_2 += 3 + val[them]
        elif b == 'X':
            score_2 += val[defeats[them]]
        else:
            score_2 += 6 + val[defeated_by[them]]

print(score, score_2)
